### ISSUE TITLE
Improve `useOthers()` for >90% use cases

### ISSUE THREAD

**Issue by nvie** [liveblocks/liveblocks#463]
The goal of this issue is to fix some drawbacks of the recently-implemented selector-based `useOthers()` hook.

It was implemented to be consistent with its current (0.17) implementation, and to be symmetric to the way `useStorage()` and `useSelf()` works.

For example:

```tsx
useStorage(root => root.some.thing)
useSelf(me => me.presence.cursor)
useOthers(others => others.map(...), shallow);
```

Since `useOthers()` is the only one of these hooks that operates on a _list_, the only reasonable operations you can do on those that make sense are:

- Some form of **mapping** - e.g. selecting all cursors
- Some form of **filtering** - e.g. selecting only users with an avatar
- Some form of **folding** - e.g. testing if _some_ or _every_ user matches a predicate

Let's consider a use case for each.

### Use cases for mapping
There are legion of these. In fact, >90% of use cases will be some kind of mapping. For example, extracting cursors, drafts, selections, avatars, etc for each of these users would require a mapping.

```tsx
// 0.18.0-beta0
const cursors = useOthers(others => others.map(other => other.presence.cursor), shallow);
const drafts = useOthers(others => others.map(other => other.presence.pencilDraft), shallow);
const avatars = useOthers(others => others.map(other => other.info.avatar), shallow);
```

It's pretty verbose, right? Also, the `shallow` here is used to compare the outer selection, which in the case of a `.map()` will always be the outer array. Not using a `shallow` comparison makes absolutely no sense with this API, and will likely always be a bug.

While selecting "stable" references means the `shallow` comparison makes sense here, it becomes problematic if we select derived data, for example, when we want to select multiple things:

```tsx
// 0.18.0-beta0
const drafts = useOthers(
  others => others.map(other => ({
    path: other.presence.pencilDraft,
    color: other.presence.penColor,
  })),
  shallow
);
```

Now, the `shallow` won't behave like you expect, because it will be used on the outermost array, but the returned `{ path, color }` objects will be `===`-compared. Because they are always new objects, these will never match. You can only solve this by using a two-level-deep-shallow comparison function 😢  Let's call this the TwoLevelShallow Comparison Problem™.

Lastly, in 99% of use cases when using `useOthers()` to get a list-of-whatever, you'll want to iterate over that list and display some UI for each item. That requires using a `key={...}` prop in React. What key to use? In pretty much all use cases, you'll want `other.connectionId`. That means that in pretty much every use case of `useOthers()`, you'll need to also select `connectionId` to retain it. Then you're back to the TwoLevelShallow Comparison Problem™.

### Use cases for filtering & folding
When filtering and folding others, you are narrowing down the list. For example:

```tsx
// In 0.18.0-beta0
const typingUsers = useOthers(others => others.filter(other => other.presence.isTyping), shallow);
const isSomeoneTyping = useOthers(others => others.some(other => other.presence.isTyping));
```

In the first case, you're using a `.filter()`, so the result needs a `shallow` check. In the second case, the result is folded into a `boolean`, so that's unnecessary.

In the `.filter()` case, though, the result will change any time the presence for _any_ of the typing users change. So the component will re-render even if one of those typing users move their cursors. In practice, you'll often want to narrow down and only select properties you want to watch for.

## A pragmatic proposal
Below, I'm proposing an alternative API for `useOthers()` that may initially look a bit weird or unusual, but that will be much nicer to use in most practical use cases.

The proposed changes:
- The 1st argument becomes an "item selector", and the final selection is always equal to `others.map(itemSelector)`. The type of that item selector is `(other: User<P, M>) => T`.
- The 2nd argument becomes an "item level comparison", and used to define how items are compared pairwise.
- The `connectionId` is always selected to make it easier to use with React use cases. The final result of the hook is always of this shape: `{ connectionId: number, data: T }[]`.
- The hook's internals ensure that comparison happens shallowly at the outer array level, and that only the `data` fields get pairwise-compared. The `{ connectionId, data }` objects always remain stable references this way.

```tsx
// Proposed to include in 0.18.0-beta1
const cursors = useOthers(other => other.presence.cursor);
cursors.map(c => <Cursor key={c.connectionId} x={c.data.x} y={c.data.y} />);

const drafts = useOthers(
  other => ({ path: other.presence.pencilDraft, color: other.presence.penColor }),
  shallow,
);
drafts.map(d => <Draft key={d.connectionId} color={d.data.color} path={d.data.path} />);

const avatars = useOthers(other => other.info.avatar);
avatars.map(a => <img src={a.data.url} />);
```

This makes it super ergonomic for the majority of use cases.

It unfortunately means that the "filtering" and folding are no longer supported use cases, though. However, those could be relatively easily expressed in terms of a map with some post-filtering, e.g.

```tsx
const users = useOthers(other => other.presence.isTyping ? other.presence.draftText : null);
const typingUsers = users.filter(u => u.data !== null);
const isSomeoneTyping = users.some(u => u.data !== null);
```

A few downsides to this API:
- May not be obvious at first (but hopefully appreciated later / in practice!)
- Is not symmetric to `useSelf()` or `useStorage()`
- May not initially be intuitive how the isEqual function applies to the item level
- The `{ connectionId, data }` return shape may feel a bit awkward at first, even though very pragmatic.
- Still, I kinda dislike the extra `.data` path needed… ideas?
- This proposal makes `useOthers()` (without selector) and `useOthers(x => x)` return different results, unlike before. Now, the second would have that `data` subfield, but the former does not 🙈

**Comment by GuillaumeSalles** [comment]
Thank you for writing up this proposal after our chat! 

I'm convinced this is more useful than the `0.18.0-beta0` API. I like the `itemSelector` to avoid the `TwoLevelShallow Comparison Problem™` :smile: 

Still unsure about the return shape and I don't see an obvious solution the last 3 downsides 🤔  Maybe an `Array<[connectionId: number, data: T]>` could be an alternative API to avoid the `data` property? Not sure it's better.

Another use case that came to my mind (that might be a bit far fetched):
Folding based on `user.id` instead of `user.connectionId` . Google docs `Avatars` component merge every "other" that share the same user id (a user that is connected on multiple tabs / device appears only once in the avatar stack)

Grouping (variation of folding) could also be useful for avatars: editors vs viewers, active vs idle.

The alternative approach that I'm still considering is the one that we briefly talked about during our call. If we assume that developers that want to use a "selector" API do it to avoid re-render and performance, we could keep the current `0.18.0-beta0` `useOthers` API (predictable and useful for folding scenarios) and propose 2 additional hooks every time "mapping" is necessary.

- `useOthersConnectionIds()` : that would be a tiny wrapper to `useOthers(others => others.map(other => other.connectionId), shallow)` which could be optimized internally by using room.suscribe("others") event (not documented). Not sure if this level of optimization is even necessary with our current number of users.
- `useOther(connectionId, other => other.presence.cursor)` also optimizable using the same event.


I'm still not completely sure about all this and I would love to get the feedback from our team and ideally from external developers. Maybe after migrating all our examples a pattern will emerge? I have a feeling like we're getting close to something but we're not fully there yet. 🤔

**Comment by nvie** [comment]
> I'm convinced this is more useful than the `0.18.0-beta0` API. I like the `itemSelector` to avoid the `TwoLevelShallow Comparison Problem™` 😄

Nice!

> Still unsure about the return shape and I don't see an obvious solution the last 3 downsides 🤔 Maybe an `Array<[connectionId: number, data: T]>` could be an alternative API to avoid the `data` property? Not sure it's better.

Yeah, I agree. That behavior is too magical to be the default. It should definitely be an opt-in. Options:

1. Adding an arg `useOthers(other => other.presence.cursor, { withIds: true })`. Bleh.
2. Adding another hook `useOthersWithId(other => other.presence.cursor)`. Hmm.

> Another use case that came to my mind (that might be a bit far fetched): Folding based on `user.id` instead of `user.connectionId` . Google docs `Avatars` component merge every "other" that share the same user id (a user that is connected on multiple tabs / device appears only once in the avatar stack)

Not far-fetched at all. I didn't explicitly call out flatMapping and/or grouping, but that would be perfectly legit and could be considered another form of folding. We would definitely lose support for doing that, which seems bad.

> `useOthersConnectionIds()` […]

I don't like this long name much, but it triggered another idea! Perhaps if we shorten the name, we could kill two birds with one stone, and offer:

- `useOtherIds()` (returning a `number[]`)
- `useOtherIds(itemSelector)` (returning a `[number, T][]`, or a `{ connectionId: number, selection: T }[]`?

If we do that, we can keep the 0.18.0-beta0 `useOthers()` implementation around separately. (But I still think using that API without _at least_ a `shallow` 2nd arg is a bug. That part still feels like a footgun to me.)

> `useOther(connectionId, other => other.presence.cursor)`

Still like this one. Will be instrumental to building performant cursor renders 👍

**Comment by nvie** [comment]
> But I still think using that API without _at least_ a `shallow` 2nd arg is a bug. That part still feels like a footgun to me.

Actually… is it 🤔 ? The use cases of `.filter()`'ing with `shallow` is perfectly safe! Folding to a boolean is also fine! Grouping would be less trivial, but still possible if you provide your own comparison function. You're in non-beginner land than anyway.

I think it's only a footgun in `.map()` situations (because of the TwoLevelShallow Comparison Problem™). But for all `.map()`-based use cases, the `useOtherIds()` seems like the more idiomatic approach anyway, because you'll always want to select the keys with those! 😄

**Comment by nvie** [comment]
To summarize, I think our best shot so far is:

- Keep `useOthers()` as-is in beta0
- Offer a new `useOtherIds(selector?, isEqual?)` API, like proposed [above](https://github.com/liveblocks/liveblocks/issues/463#issuecomment-1239131308)
- Offer a new (singular) `useOther(connectionId, selector?)` API
- Get more real-world feedback on those!

**Comment by nvie** [comment]
Closing for now, as this proposal has landed in 0.18.0-beta1. A final name tweak may or may not still land before the final launch, but I'm happy with the conceptual addition here, as it works well in practice.
